// rsuite-datepicker/lib/Toolbar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _rsuiteIntl = require('rsuite-intl');

var _utils = require('rsuite-utils/lib/utils');

var _Picker = require('rsuite-utils/lib/Picker');

var _isOneOf = require('./utils/isOneOf');

var _isOneOf2 = _interopRequireDefault(_isOneOf);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var namespace = _Picker.constants.namespace;

var Toolbar = function (_React$PureComponent) {
  _inherits(Toolbar, _React$PureComponent);

  function Toolbar() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Toolbar);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Toolbar.__proto__ || Object.getPrototypeOf(Toolbar)).call.apply(_ref, [this].concat(args))), _this), _this.addPrefix = function (name) {
      return (0, _utils.prefix)(_this.props.classPrefix)(name);
    }, _this.hasLocaleKey = function (key) {
      var ranges = _this.props.ranges;

      var keys = ranges.map(function (item) {
        return item.label;
      });
      return (0, _isOneOf2.default)(key, keys);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Toolbar, [{
    key: 'renderOkButton',
    value: function renderOkButton() {
      var _props = this.props,
          disabledHandle = _props.disabledHandle,
          pageDate = _props.pageDate,
          onOk = _props.onOk;


      var disabled = disabledHandle && disabledHandle(pageDate);
      var classes = (0, _classnames2.default)(this.addPrefix('right-btn-ok'), _defineProperty({}, this.addPrefix('btn-disabled'), disabled));
      return React.createElement(
        'div',
        { className: this.addPrefix('right') },
        React.createElement(
          'button',
          { className: classes, onClick: disabled ? undefined : onOk },
          React.createElement(_rsuiteIntl.FormattedMessage, { id: 'ok' })
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props2 = this.props,
          ranges = _props2.ranges,
          onShortcut = _props2.onShortcut,
          disabledHandle = _props2.disabledHandle,
          className = _props2.className,
          pageDate = _props2.pageDate,
          classPrefix = _props2.classPrefix,
          rest = _objectWithoutProperties(_props2, ['ranges', 'onShortcut', 'disabledHandle', 'className', 'pageDate', 'classPrefix']);

      var classes = (0, _classnames2.default)(classPrefix, className);
      var unhandled = (0, _utils.getUnhandledProps)(Toolbar, rest);

      return React.createElement(
        'div',
        _extends({}, unhandled, { className: classes }),
        React.createElement(
          'div',
          { className: this.addPrefix('ranges') },
          ranges.map(function (item, index) {
            var value = typeof item.value === 'function' ? item.value(pageDate) : item.value;
            var disabled = disabledHandle && disabledHandle(value);
            var itemClassName = (0, _classnames2.default)(_this2.addPrefix('option'), _defineProperty({}, _this2.addPrefix('option-disabled'), disabled));
            return React.createElement(
              'a',
              {
                /* eslint-disable */
                key: index,
                role: 'button',
                tabIndex: '-1',
                className: itemClassName,
                onClick: function onClick(event) {
                  !disabled && onShortcut && onShortcut(value, item.closeOverlay, event);
                }
              },
              _this2.hasLocaleKey(item.label) ? React.createElement(_rsuiteIntl.FormattedMessage, { id: item.label }) : item.label
            );
          })
        ),
        this.renderOkButton()
      );
    }
  }]);

  return Toolbar;
}(React.PureComponent);

Toolbar.defaultProps = {
  classPrefix: namespace + '-toolbar',
  ranges: [{
    label: 'today',
    value: (0, _moment2.default)(),
    closeOverlay: true
  }, {
    label: 'yesterday',
    value: (0, _moment2.default)().add(-1, 'd'),
    closeOverlay: true
  }]
};
Toolbar.handledProps = ['className', 'classPrefix', 'disabledHandle', 'onOk', 'onShortcut', 'pageDate', 'ranges'];
Toolbar.propTypes = {
  ranges: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    label: _propTypes2.default.node.isRequired,
    closeOverlay: _propTypes2.default.bool,
    value: _propTypes2.default.oneOfType([function () {
      return (typeof moment$Moment === 'function' ? _propTypes2.default.instanceOf(moment$Moment) : _propTypes2.default.any).apply(this, arguments);
    }, _propTypes2.default.func]).isRequired
  })).isRequired,
  className: _propTypes2.default.string,
  classPrefix: _propTypes2.default.string,
  pageDate: function pageDate() {
    return (typeof moment$Moment === 'function' ? _propTypes2.default.instanceOf(moment$Moment) : _propTypes2.default.any).apply(this, arguments);
  },
  onShortcut: _propTypes2.default.func,
  onOk: _propTypes2.default.func,
  disabledHandle: _propTypes2.default.func
};
exports.default = Toolbar;
